modules/transactionpool/accept.go

```diff
@@ -0,0 +1,182 @@
+			continue
+		}
+
+		err = errors.New("invalid input in transaction")
+		return
+	}
+
+	return
+}
+
+// validTransaction checks that there are no double spends and that all other
+// parts of the transaction are legal according to the Standard.md rules and
+// the consensus rules.
+func (tp *TransactionPool) validTransaction(t consensus.Transaction) (err error) {
+	// Get the input sum and verify that all inputs come from a valid source
+	// (confirmed or unconfirmed).
+	inputSum, err := tp.checkInputs(t)
+	if err != nil {
+		return
+	}
+
+	// Check that the inputs equal the outputs.
+	if inputSum != t.OutputSum() {
+		err = errors.New("input sum does not equal output sum")
+		return
+	}
+
+	// Check that the signatures are valid.
+	err = tp.state.ValidSignatures(t)
+	if err != nil {
+		return
+	}
+
+	// Check that all contracts are legal within the existing state. This means
+	// that they have a non-zero payout and that the start date is less than
+	// the current height of the blockchain.
+	for _, contract := range t.FileContracts {
+		err = tp.state.ValidContract(contract)
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+// addTransaction adds a transaction to the transaction pool.
+func (tp *TransactionPool) addTransaction(t consensus.Transaction) {
+	ut := &unconfirmedTransaction{
+		transaction: t,
+	}
+
+	// Go through the inputs and them to the used outputs list, updating the
+	// requirements and dependents as necessary.
+	for _, input := range t.Inputs {
+		// Sanity check - this input should not already be in the usedOutputs
+		// list.
+		if consensus.DEBUG {
+			_, exists := tp.usedOutputs[input.OutputID]
+			if exists {
+				panic("addTransaction called on invalid transaction")
+			}
+		}
+
+		unconfirmedTxn, exists := tp.newOutputs[input.OutputID]
+		if exists {
+			unconfirmedTxn.dependents = append(unconfirmedTxn.dependents, ut)
+			ut.requirements = append(ut.requirements, unconfirmedTxn)
+		}
+		tp.usedOutputs[input.OutputID] = ut
+	}
+
+	// Add each new output to the list of outputs and newOutputs.
+	for i, output := range t.Outputs {
+		// Sanity check - this output should not already exist in newOutputs or
+		// outputs.
+		if consensus.DEBUG {
+			_, exists := tp.newOutputs[t.OutputID(i)]
+			if exists {
+				panic("trying to add an output that already exists?")
+			}
+			_, exists = tp.outputs[t.OutputID(i)]
+			if exists {
+				panic("trying to add an output that already exists?")
+			}
+		}
+
+		tp.outputs[t.OutputID(i)] = output
+		tp.newOutputs[t.OutputID(i)] = ut
+	}
+
+	tp.addTransactionToTail(ut)
+}
+
+// AcceptTransaction takes a new transaction from the network and puts it in
+// the transaction pool after checking it for legality and consistency.
+func (tp *TransactionPool) AcceptTransaction(t consensus.Transaction) (err error) {
+	tp.mu.Lock()
+	defer tp.mu.Unlock()
+	tp.state.RLock()
+	defer tp.state.RUnlock()
+
+	// Check that the transaction follows 'Standard.md' guidelines.
+	err = standard(t)
```

Comment:
hmmm. Bitcoin has an equivalent function named IsStandard(). would that be a better name?

It basically enforces a set of rules that protects miners against potential future soft forks without requiring the miners to upgrade when the soft forks happen. It's a good thing to have around.